[lldb] Change Module to have a concrete UnwindTable, update

jasonmolenda opened the pull request:
Currently a Module has a std::optional<UnwindTable> which is created when the UnwindTable is requested from outside the Module.  The idea is to delay its creation until the Module has an ObjectFile initialized, which will have been done by the time we're doing an unwind.

However, Module::GetUnwindTable wasn't doing any locking, so it was possible for two threads to ask for the UnwindTable for the first time, one would be created and returned while another thread would create one, destroy the first in the process of emplacing it. It was an uncommon crash, but it was possible.

Grabbing the Module's mutex would be one way to address it, but when loading ELF binaries, we start creating the SymbolTable on one thread (ObjectFileELF) grabbing the Module's mutex, and then spin up worker threads to parse the individual DWARF compilation units, which then try to also get the UnwindTable and deadlock if they try to get the Module's mutex.

This changes Module to have a concrete UnwindTable as an ivar, and when it adds an ObjectFile or SymbolFileVendor, it will call the Update method on it, which will re-evaluate which sections exist in the ObjectFile/SymbolFile. UnwindTable used to have an Initialize method which set all the sections, and an Update method which would set some of them if they weren't set.  I unified these with the Initialize method taking a `force` option to re-initialize the section pointers even if they had been done already before.

This is addressing a rare crash report we've received, and also a failure Adrian spotted on the -fsanitize=address CI bot last week, it's still uncommon with ASAN but it can happen with the standard testsuite.

rdar://128876433

jasonmolenda committed:
[lldb] Change Module to have a concrete UnwindTable, update

Currently a Module has a std::optional<UnwindTable> which is
created when the UnwindTable is requested from outside the
Module.  The idea is to delay its creation until the Module
has an ObjectFile initialized, which will have been done by
the time we're doing an unwind.

However, Module::GetUnwindTable wasn't doing any locking, so it was
possible for two threads to ask for the UnwindTable for the first
time, one would be created and returned while another thread would
create one, destroy the first in the process of emplacing it.
It was an uncommon crash, but it was possible.

Grabbing the Module's mutex would be one way to address it, but
when loading ELF binaries, we start creating the SymbolTable on one
thread (ObjectFileELF) grabbing the Module's mutex, and then spin
up worker threads to parse the individual DWARF compilation units,
which then try to also get the UnwindTable and deadlock if they try
to get the Module's mutex.

This changes Module to have a concrete UnwindTable as an ivar, and
when it adds an ObjectFile or SymbolFileVendor, it will call the
Update method on it, which will re-evaluate which sections exist
in the ObjectFile/SymbolFile. UnwindTable used to have an Initialize
method which set all the sections, and an Update method which would
set some of them if they weren't set.  I unified these with the
Initialize method taking a `force` option to re-initialize the
section pointers even if they had been done already before.

This is addressing a rare crash report we've received, and also a
failure Adrian spotted on the -fsanitize=address CI bot last week,
it's still uncommon with ASAN but it can happen with the standard
testsuite.

rdar://132514736

jasonmolenda committed:
Only recalculate the UnwindTable input sources if they are
unset.

jimingham commented on lldb/include/lldb/Symbol/UnwindTable.h:
```diff
@@ -64,7 +64,7 @@ class UnwindTable {
   void Dump(Stream &s);
 
-  void Initialize();
+  void Initialize(bool force = false);
```
This needs a comment explaining what you are forcing.

jasonmolenda replied:
Yeah I'm not thrilled with the terminology I used.  But I had `Initialize` which assumed no unwind sources had been discovered, and `Update` called after a SymbolFileVendor is added to the Module which looks for any un-set unwind sources and checks to see if they can now be found.  That duplication didn't thrill me.

jimingham commented:
So the strategy here is the UnwindTable class calls its `Initialize()` any time a client wants to use an UnwindTable so that it can be lazily filled.   Then Module calls `UnwindTable::Update()` any time something gets added to the module that might change the unwind table.

That seems pretty clear.  

It is a little odd that the laziness of the UnwindTable itself gets defeated by Update, which forces the Initialize even if the unwind table isn't currently being requested.

Not sure if that matters?  If it does, Update could just tell the UnwindTable that next time it gets asked a question it has to reread the unwind information.  Maybe it could even just set m_initialized back to `false` and  then let the lazy initialization redo the work when requested?

jasonmolenda commented:
> It is a little odd that the laziness of the UnwindTable itself gets defeated by Update, which forces the Initialize even if the unwind table isn't currently being requested.

I could add an ivar `m_module_has_updated` which is set when a new ObjectFile/SymbolFileVendor is added to the Module.  And all the getter methods in UnwindTable call `Initialize()` which currently checks that it has been initialized or not, and add that in.

> Not sure if that matters? If it does, Update could just tell the UnwindTable that next time it gets asked a question it has to reread the unwind information. Maybe it could even just set m_initialized back to `false` and then let the lazy initialization redo the work when requested?

Hah, I didn't read your full comment.  Yes, same idea, and just as good.  I'll do that.

jasonmolenda committed:
Don't non-lazily parse unwind infos when SymbolFile added

Rename UnwindTable::Update to UnwindTable::ModuleWasUpdated.

Remove `force` parameter to `Initialize` method.  Have
`ModuleWasUpdated` unset m_initialized, so the next time
the UnwindTable is asked for a source of information, we
will then parse any newly available sources of unwind
information.

jasonmolenda commented:
Thanks for the suggestion @jimingham , I pushed an update to implement that idea.

JDevlieghere reviewed (approved):
This makes sense to me. LGTM if Jim is happy.

jimingham reviewed (approved):
LGTM I like that you also changed "Update" to "ModuleWasUpdated", that's more accurate.  `m_initialized` isn't quite the right name, it means "needs work" not "initialized" which sounds like it should happen once.
If you can think of a better name for that off-hand, that would be a little clearer, but I can't think of a good one, and it's not crucial.

jasonmolenda committed:
rename m_initialized to m_scanned_all_unwind_sources
in UnwindTable class, to make it clearer what it is tracking.
